File: tennis-server/src/main/java/com/tennis/service/MatchService.java
package com.tennis.service;

import java.io.IOException;
import java.io.OutputStream;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import org.apache.poi.ss.usermodel.Row;
import org.apache.poi.ss.usermodel.Sheet;
import org.apache.poi.ss.usermodel.Workbook;
import org.apache.poi.xssf.usermodel.XSSFWorkbook;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.tennis.dto.MatchCurrentScoreDTO;
import com.tennis.dto.MatchStartRequest;
import com.tennis.dto.MatchStatsDTO;
import com.tennis.entity.Match;
import com.tennis.entity.Player;
import com.tennis.entity.Rally;
import com.tennis.repository.MatchRepository;
import com.tennis.repository.PlayerRepository;
import com.tennis.repository.RallyRepository;

@Service
public class MatchService {
    @Autowired
    private MatchRepository matchRepository;
    @Autowired
    private RallyRepository rallyRepository;
    @Autowired
    private PlayerRepository playerRepository;

    public Match startMatch(MatchStartRequest request) {
        Match match = new Match();
        match.setDate(LocalDate.now());
        match.setPlayer1Id(request.getPlayer1Id());
        match.setPlayer2Id(request.getPlayer2Id());
        match.setStatus("started");
        match.setStartedAt(LocalDateTime.now());
        match.setFirstServer(request.getFirstServer());
        return matchRepository.save(match);
    }

    public Match finishMatch(Long id) {
        Match match = matchRepository.findById(id).orElseThrow(() -> new RuntimeException("Match not found"));
        match.setStatus("finished");
        return matchRepository.save(match);
    }

    public MatchCurrentScoreDTO getCurrentScore(Long id) {
        List<Rally> rallies = rallyRepository.findByMatchId(id);
        int scoreP1 = 0;
        int scoreP2 = 0;
        int rallyNumber = 0;

        if (!rallies.isEmpty()) {
            rallies.sort(Comparator.comparingInt(Rally::getRallyNumber).reversed());
            Rally lastRally = rallies.get(0);
            if (lastRally.getScoreP1() != null) scoreP1 = lastRally.getScoreP1();
            if (lastRally.getScoreP2() != null) scoreP2 = lastRally.getScoreP2();
            rallyNumber = lastRally.getRallyNumber();
        }

        // Calculate current server based on total score
        String server = "Unknown";
        Match match = matchRepository.findById(id).orElse(null);
        
        if (match != null && match.getFirstServer() != null) {
            Long firstServerId = match.getFirstServer();
            Long p1Id = match.getPlayer1Id();
            Long p2Id = match.getPlayer2Id();
            
            int totalScore = scoreP1 + scoreP2;
            int turns;
            
            // Logic: Switch every 2 points until both reach 10 (Deuce), then switch every 1 point
            if (scoreP1 >= 10 && scoreP2 >= 10) {
                // Deuce logic: 10-10 means 20 points played.
                // First 20 points (0-19) were 10 turns (2 serves each).
                // Points from 20 onwards are 1 serve each.
                turns = 10 + (totalScore - 20);
            } else {
                // Normal logic: Switch every 2 points
                turns = totalScore / 2;
            }
            
            Long currentServerId;
            if (turns % 2 == 0) {
                currentServerId = firstServerId;
            } else {
                currentServerId = firstServerId.equals(p1Id) ? p2Id : p1Id;
            }
            
            if (currentServerId.equals(p1Id)) {
                server = "player1";
            } else {
                server = "player2";
            }
        }

        MatchCurrentScoreDTO dto = new MatchCurrentScoreDTO(id, rallyNumber, scoreP1, scoreP2, server);
        
        // Populate player names
        if (match != null) {
            playerRepository.findById(match.getPlayer1Id()).ifPresent(p -> dto.setPlayer1Name(p.getName()));
            playerRepository.findById(match.getPlayer2Id()).ifPresent(p -> dto.setPlayer2Name(p.getName()));
        }
        
        return dto;
    }

    public List<Match> getMatchHistory(LocalDate dateFrom, LocalDate dateTo, String playerName, String status) {
        List<Match> allMatches = matchRepository.findAll();
        
        // Simple in-memory filtering
        return allMatches.stream()
                .filter(m -> dateFrom == null || !m.getDate().isBefore(dateFrom))
                .filter(m -> dateTo == null || !m.getDate().isAfter(dateTo))
                .filter(m -> status == null || status.isEmpty() || m.getStatus().equalsIgnoreCase(status))
                .filter(m -> {
                    if (playerName == null || playerName.isEmpty()) return true;
                    // Find player name in either player1 or player2
                    Player p1 = playerRepository.findById(m.getPlayer1Id()).orElse(null);
                    Player p2 = playerRepository.findById(m.getPlayer2Id()).orElse(null);
                    
                    boolean p1Match = p1 != null && p1.getName().contains(playerName);
                    boolean p2Match = p2 != null && p2.getName().contains(playerName);
                    
                    return p1Match || p2Match;
                })
                .peek(m -> {
                    playerRepository.findById(m.getPlayer1Id()).ifPresent(p -> m.setPlayer1Name(p.getName()));
                    playerRepository.findById(m.getPlayer2Id()).ifPresent(p -> m.setPlayer2Name(p.getName()));
                })
                .collect(Collectors.toList());
    }

    public List<Rally> getMatchDetail(Long id) {
        return rallyRepository.findByMatchId(id);
    }

    public MatchStatsDTO getMatchStats(Long id) {
        List<Rally> rallies = rallyRepository.findByMatchId(id);
        MatchStatsDTO stats = new MatchStatsDTO();
        
        // Calculate stats
        Map<String, Double> serveSuccessRate = new HashMap<>();
        // Placeholder logic: assume 100% or calculate if we had "faults"
        // But we only have "scorer". 
        // If server == scorer, then serve win?
        long p1ServeCount = rallies.stream().filter(r -> "player1".equalsIgnoreCase(r.getServer())).count();
        long p1ServeWin = rallies.stream().filter(r -> "player1".equalsIgnoreCase(r.getServer()) && "player1".equalsIgnoreCase(r.getScorer())).count();
        
        long p2ServeCount = rallies.stream().filter(r -> "player2".equalsIgnoreCase(r.getServer())).count();
        long p2ServeWin = rallies.stream().filter(r -> "player2".equalsIgnoreCase(r.getServer()) && "player2".equalsIgnoreCase(r.getScorer())).count();

        serveSuccessRate.put("player1", p1ServeCount > 0 ? (double)p1ServeWin / p1ServeCount : 0.0);
        serveSuccessRate.put("player2", p2ServeCount > 0 ? (double)p2ServeWin / p2ServeCount : 0.0);
        stats.setServeSuccessRate(serveSuccessRate);

        Map<String, Integer> consecutiveScore = new HashMap<>();
        // Calculate max consecutive score
        consecutiveScore.put("player1", calculateMaxConsecutive(rallies, "player1"));
        consecutiveScore.put("player2", calculateMaxConsecutive(rallies, "player2"));
        stats.setConsecutiveScore(consecutiveScore);

        stats.setAverageRallyTime(15.2); // Mock or calculate if we had duration

        return stats;
    }

    private int calculateMaxConsecutive(List<Rally> rallies, String player) {
        int max = 0;
        int current = 0;
        // Sort by rally number
        rallies.sort(Comparator.comparingInt(Rally::getRallyNumber));
        
        for (Rally r : rallies) {
            if (player.equalsIgnoreCase(r.getScorer())) {
                current++;
            } else {
                max = Math.max(max, current);
                current = 0;
            }
        }
        return Math.max(max, current);
    }

    public void exportMatch(Long id, OutputStream out) throws IOException {
        Match match = matchRepository.findById(id).orElseThrow(() -> new RuntimeException("Match not found"));
        List<Rally> rallies = rallyRepository.findByMatchId(id);
        rallies.sort(Comparator.comparingInt(Rally::getRallyNumber));

        // Get Player Names
        String player1Name = playerRepository.findById(match.getPlayer1Id()).map(Player::getName).orElse("Unknown");
        String player2Name = playerRepository.findById(match.getPlayer2Id()).map(Player::getName).orElse("Unknown");

        try (Workbook workbook = new XSSFWorkbook()) {
            Sheet sheet = workbook.createSheet("比赛数据");
            
            // Header
            Row headerRow = sheet.createRow(0);
            String[] columns = {"回合数", "得分者", "比分 (P1-P2)", "发球方", "时间"};
            for (int i = 0; i < columns.length; i++) {
                headerRow.createCell(i).setCellValue(columns[i]);
            }

            // Date Formatter
            java.time.format.DateTimeFormatter formatter = java.time.format.DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");

            // Data
            int rowNum = 1;
            for (Rally rally : rallies) {
                Row row = sheet.createRow(rowNum++);
                row.createCell(0).setCellValue(rally.getRallyNumber());
                
                // Map scorer to name
                String scorerName = "player1".equalsIgnoreCase(rally.getScorer()) ? player1Name : 
                                   ("player2".equalsIgnoreCase(rally.getScorer()) ? player2Name : rally.getScorer());
                row.createCell(1).setCellValue(scorerName);
                
                row.createCell(2).setCellValue(rally.getScoreP1() + "-" + rally.getScoreP2());
                
                // Map server to name
                String serverName = "player1".equalsIgnoreCase(rally.getServer()) ? player1Name : 
                                   ("player2".equalsIgnoreCase(rally.getServer()) ? player2Name : rally.getServer());
                row.createCell(3).setCellValue(serverName);
                
                row.createCell(4).setCellValue(rally.getTimestamp().format(formatter));
            }

            workbook.write(out);
        }
    }
}

File: pingpong-backend/internal/service/match_service.go
package service

import (
	"errors"
	"io"
	"sort"
	"strconv"
	"strings"
	"time"

	"pingpong-backend/internal/dto"
	"pingpong-backend/internal/model"

	"github.com/xuri/excelize/v2"
	"gorm.io/gorm"
)

type MatchService struct {
	db *gorm.DB
}

func NewMatchService(db *gorm.DB) *MatchService {
	return &MatchService{db: db}
}

func (s *MatchService) StartMatch(request dto.MatchStartRequest) (model.Match, error) {
	match := model.Match{
		Date:        model.TodayDate(),
		Player1ID:   request.Player1ID,
		Player2ID:   request.Player2ID,
		Status:      "started",
		StartedAt:   time.Now(),
		FirstServer: request.FirstServer,
	}
	if err := s.db.Create(&match).Error; err != nil {
		return model.Match{}, err
	}
	return match, nil
}

func (s *MatchService) GetMatch(id uint64) (model.Match, error) {
	var match model.Match
	if err := s.db.First(&match, id).Error; err != nil {
		return model.Match{}, err
	}
	match.Player1Name, match.Player2Name = s.playerNames(match)
	return match, nil
}

func (s *MatchService) FinishMatch(id uint64) (model.Match, error) {
	var match model.Match
	if err := s.db.First(&match, id).Error; err != nil {
		return model.Match{}, err
	}
	match.Status = "finished"
	if err := s.db.Save(&match).Error; err != nil {
		return model.Match{}, err
	}
	return match, nil
}

func (s *MatchService) CurrentScore(id uint64) (dto.MatchCurrentScoreResponse, error) {
	var rallies []model.Rally
	if err := s.db.Where("match_id = ?", id).Find(&rallies).Error; err != nil {
		return dto.MatchCurrentScoreResponse{}, err
	}

	scoreP1, scoreP2 := LatestScore(rallies)
	rallyNumber := 0
	if len(rallies) > 0 {
		rallyNumber = LatestRally(rallies).RallyNumber
	}

	var match model.Match
	err := s.db.First(&match, id).Error
	if errors.Is(err, gorm.ErrRecordNotFound) {
		return dto.MatchCurrentScoreResponse{}, err
	}
	if err != nil {
		return dto.MatchCurrentScoreResponse{}, err
	}

	p1ID := match.Player1ID
	p2ID := match.Player2ID
	response := dto.MatchCurrentScoreResponse{
		MatchID:     id,
		RallyNumber: rallyNumber,
		ScoreP1:     scoreP1,
		ScoreP2:     scoreP2,
		Server:      CalculateServer(scoreP1, scoreP2, match.FirstServer, &p1ID, &p2ID),
	}
	response.Player1Name, response.Player2Name = s.playerNames(match)
	return response, nil
}

func (s *MatchService) MatchHistory(dateFrom, dateTo *time.Time, playerName, status string) ([]model.Match, error) {
	var matches []model.Match
	query := s.db.Model(&model.Match{})
	if dateFrom != nil {
		query = query.Where("date >= ?", dateFrom.Format("2006-01-02"))
	}
	if dateTo != nil {
		query = query.Where("date <= ?", dateTo.Format("2006-01-02"))
	}
	if status != "" {
		query = query.Where("LOWER(status) = LOWER(?)", status)
	}
	if playerName != "" {
		query = query.Joins("LEFT JOIN player p1 ON p1.id = matches.player1_id").
			Joins("LEFT JOIN player p2 ON p2.id = matches.player2_id").
			Where("p1.name LIKE ? OR p2.name LIKE ?", "%"+playerName+"%", "%"+playerName+"%")
	}
	if err := query.Find(&matches).Error; err != nil {
		return nil, err
	}

	for i := range matches {
		matches[i].Player1Name, matches[i].Player2Name = s.playerNames(matches[i])
	}
	return matches, nil
}

func (s *MatchService) MatchDetail(id uint64) ([]model.Rally, error) {
	var rallies []model.Rally
	err := s.db.Where("match_id = ?", id).Order("rally_number ASC").Find(&rallies).Error
	return rallies, err
}

func (s *MatchService) MatchStats(id uint64) (dto.MatchStatsResponse, error) {
	var rallies []model.Rally
	if err := s.db.Where("match_id = ?", id).Find(&rallies).Error; err != nil {
		return dto.MatchStatsResponse{}, err
	}

	p1ServeCount, p1ServeWin := 0, 0
	p2ServeCount, p2ServeWin := 0, 0
	for _, rally := range rallies {
		if strings.EqualFold(rally.Server, "player1") {
			p1ServeCount++
			if strings.EqualFold(rally.Scorer, "player1") {
				p1ServeWin++
			}
		}
		if strings.EqualFold(rally.Server, "player2") {
			p2ServeCount++
			if strings.EqualFold(rally.Scorer, "player2") {
				p2ServeWin++
			}
		}
	}

	return dto.MatchStatsResponse{
		ServeSuccessRate: map[string]float64{
			"player1": rate(p1ServeWin, p1ServeCount),
			"player2": rate(p2ServeWin, p2ServeCount),
		},
		ConsecutiveScore: map[string]int{
			"player1": MaxConsecutiveScore(append([]model.Rally(nil), rallies...), "player1"),
			"player2": MaxConsecutiveScore(append([]model.Rally(nil), rallies...), "player2"),
		},
		AverageRallyTime: 15.2,
	}, nil
}

func (s *MatchService) ExportMatch(id uint64, out io.Writer) error {
	var match model.Match
	if err := s.db.First(&match, id).Error; err != nil {
		return err
	}
	var rallies []model.Rally
	if err := s.db.Where("match_id = ?", id).Order("rally_number ASC").Find(&rallies).Error; err != nil {
		return err
	}

	player1Name, player2Name := s.playerNames(match)
	if player1Name == "" {
		player1Name = "Unknown"
	}
	if player2Name == "" {
		player2Name = "Unknown"
	}

	file := excelize.NewFile()
	defer file.Close()

	defaultSheet := file.GetSheetName(0)
	sheet := "姣旇禌鏁版嵁"
	if err := file.SetSheetName(defaultSheet, sheet); err != nil {
		return err
	}

	columns := []string{"鍥炲悎鏁?", "寰楀垎鑰?", "姣斿垎 (P1-P2)", "鍙戠悆鏂?", "鏃堕棿"}
	for idx, column := range columns {
		cell, err := excelize.CoordinatesToCellName(idx+1, 1)
		if err != nil {
			return err
		}
		if err := file.SetCellValue(sheet, cell, column); err != nil {
			return err
		}
	}

	sort.Slice(rallies, func(i, j int) bool {
		return rallies[i].RallyNumber < rallies[j].RallyNumber
	})
	for idx, rally := range rallies {
		row := idx + 2
		values := []any{
			rally.RallyNumber,
			playerLabel(rally.Scorer, player1Name, player2Name),
			scoreLabel(rally.ScoreP1, rally.ScoreP2),
			playerLabel(rally.Server, player1Name, player2Name),
			rally.Timestamp.Format("2006-01-02 15:04:05"),
		}
		for col, value := range values {
			cell, err := excelize.CoordinatesToCellName(col+1, row)
			if err != nil {
				return err
			}
			if err := file.SetCellValue(sheet, cell, value); err != nil {
				return err
			}
		}
	}
	return file.Write(out)
}

func (s *MatchService) playerNames(match model.Match) (string, string) {
	var p1, p2 model.Player
	player1Name, player2Name := "", ""
	if err := s.db.First(&p1, match.Player1ID).Error; err == nil {
		player1Name = p1.Name
	}
	if err := s.db.First(&p2, match.Player2ID).Error; err == nil {
		player2Name = p2.Name
	}
	return player1Name, player2Name
}

func rate(wins, total int) float64 {
	if total == 0 {
		return 0.0
	}
	return float64(wins) / float64(total)
}

func playerLabel(value, player1Name, player2Name string) string {
	if strings.EqualFold(value, "player1") {
		return player1Name
	}
	if strings.EqualFold(value, "player2") {
		return player2Name
	}
	return value
}

func scoreLabel(scoreP1, scoreP2 *int) string {
	p1, p2 := 0, 0
	if scoreP1 != nil {
		p1 = *scoreP1
	}
	if scoreP2 != nil {
		p2 = *scoreP2
	}
	return strings.Join([]string{itoa(p1), itoa(p2)}, "-")
}

func itoa(value int) string {
	return strconv.Itoa(value)
}
